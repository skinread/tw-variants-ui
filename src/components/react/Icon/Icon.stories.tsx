import type { Meta, StoryObj } from '@storybook/react';

import { Icon as ReactIcon } from './Icon';

const meta = {
  title: 'React/Icon',
  component: ReactIcon,
  tags: ['autodocs'],
  args: {
    name: undefined,
    className: 'fill-primary stroke-primary',
    width: 100,
  },
} satisfies Meta<typeof ReactIcon>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Icon: Story = {
  args: {
    name: 'info',
  },
};
