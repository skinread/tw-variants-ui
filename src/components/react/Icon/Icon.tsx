import IconArrowBack from '@assets/icon-arrow-back.svg?react';
import IconCross from '@assets/icon-cross.svg?react';
import IconEmail from '@assets/icon-email.svg?react';
import IconExclamation from '@assets/icon-exclamation.svg?react';
import IconExternal from '@assets/icon-external.svg?react';
import IconEyeCrossed from '@assets/icon-eye-crossed.svg?react';
import IconEye from '@assets/icon-eye.svg?react';
import IconInfo from '@assets/icon-info.svg?react';
import IconTick from '@assets/icon-tick-circle.svg?react';

const iconMap = {
  'arrow-back': IconArrowBack,
  cross: IconCross,
  email: IconEmail,
  exclamation: IconExclamation,
  'tick-circle': IconTick,
  external: IconExternal,
  eye: IconEye,
  'eye-crossed': IconEyeCrossed,
  info: IconInfo,
} as const;

export type SvgProps = React.ComponentPropsWithoutRef<'svg'>;
export type AllIcons = keyof typeof iconMap;
export type FeedbackIcons = Extract<AllIcons, 'info' | 'exclamation' | 'email' | 'tick-circle'>;
export type NavigationIcons = Extract<AllIcons, 'arrow-back' | 'info'>;
export interface IconProps extends SvgProps {
  /**
   * shortcut reference name of icon to display
   */
  name: AllIcons;
  className?: SvgProps['className'];
}

/**
 * Convenience component for SVG icon glyphs. Colour is set from `fill` for single colour icons. Size defaults to `1em`
 * or set by `width`/`height` attribute or `className`. Icons are just decorative, they do not have alt text like images.
 */
export const Icon = ({ className, name, ...attrs }: IconProps) => {
  const Svg = iconMap[name];

  let defaultSize = null;
  if (!className && !attrs.width && !attrs.height) defaultSize = 'w-[1em] h-[1em]';

  const svgClassName = ['inline-block', defaultSize, className, `icon-${name}`].join(' ');
  return <Svg className={svgClassName} {...attrs} aria-hidden />;
};
