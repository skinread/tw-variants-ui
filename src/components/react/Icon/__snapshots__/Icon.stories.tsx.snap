// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React/Icon Icon smoke-test 1`] = `
<svg xmlns="http://www.w3.org/2000/svg"
     viewbox="0 0 24 24"
     fill="currentColor"
     class="inline-block  fill-primary stroke-primary icon-info"
     width="100"
     aria-hidden="true"
>
  <path d="M12,24A12,12,0,1,1,24,12,12.013,12.013,0,0,1,12,24ZM12,2A10,10,0,1,0,22,12,10.011,10.011,0,0,0,12,2Z"
        stroke="none"
  >
  </path>
  <path d="M14,19H12V12H10V10h2a2,2,0,0,1,2,2Z"
        stroke="none"
  >
  </path>
  <circle cx="12"
          cy="6.5"
          r="1.5"
  >
  </circle>
</svg>
`;
